node/router/src/handshake.rs

```diff
@@ -304,22 +332,30 @@ impl<N: Network> Router<N> {
     }
 
     /// Verifies the given challenge response. Returns a disconnect reason if the response is invalid.
+    #[allow(clippy::too_many_arguments)]
     async fn verify_challenge_response(
         &self,
         peer_addr: SocketAddr,
         peer_address: Address<N>,
+        peer_node_type: NodeType,
         response: ChallengeResponse<N>,
         expected_genesis_header: Header<N>,
+        expected_restrictions_id: Field<N>,
         expected_nonce: u64,
     ) -> Option<DisconnectReason> {
         // Retrieve the components of the challenge response.
-        let ChallengeResponse { genesis_header, signature, nonce } = response;
+        let ChallengeResponse { genesis_header, restrictions_id, signature, nonce } = response;
 
         // Verify the challenge response, by checking that the block header matches.
         if genesis_header != expected_genesis_header {
             warn!("Handshake with '{peer_addr}' failed (incorrect block header)");
             return Some(DisconnectReason::InvalidChallengeResponse);
         }
+        // Verify the restrictions ID.
+        if !peer_node_type.is_prover() && restrictions_id != expected_restrictions_id {
```

Comment:
Would this still cause a prover itself to disconnect if the restrictions_id doesn't match? So shouldn't it be:

```suggestion
        if !peer_node_type.is_prover() && !self.node_type.is_prover() && restrictions_id != expected_restrictions_id {
```